### PR TITLE
Update trace propagation format for botocore direct invocations

### PR DESCRIPTION
**Pull request by agocs**
## Description

Updates the trace propagation format for botocore direct lambda invocations. The old way was:

```
clientContext: {
  custom:{
    _datadog: {
      x-datadog-trace-id: "",
      x-datadog-parent-id: "",
      ...
    }
  }
}
```

New way is:

```
clientContext: {
  custom: {
      x-datadog-trace-id: "",
      x-datadog-parent-id: "",
      ...
  }
}
```

https://docs.google.com/document/d/1pruSeesAr2bAVMtz_rUlsiBZKFoFWdB4nSq5Vuym2l8/edit?usp=sharing

### Reasoning

The Go and Java runtimes only support `ClientContext.custom` fields of type `map[string]string`. C.f. :
- https://github.com/aws/aws-lambda-java-libs/issues/212
- https://github.com/DataDog/dd-trace-js/issues/1312

In order to pass trace contexts to directly invoked Go and Java lambda functions, we can write the trace context key:values to ClientContext.custom rather than creating a ClientContext.custom._datadog object like we've been doing.

### Backwards compatibility

This is, strictly speaking, a breaking change in how we add trace context to direct invocations. The Python and JS repos have already been updated to accept the new format:
- https://github.com/DataDog/datadog-lambda-js/pull/194
- https://github.com/DataDog/datadog-lambda-python/pull/146

They both check for the old and new formats. 

## Checklist
- [x] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

**Review by Kyle-Verhoog — COMMENTED** [review]
I'm thinking we should leave the old method behind a configuration option since a bunch of users will not be on the latest version of the serverless client(s) 🤔 

Thoughts @agocs?

**Comment by agocs** [comment]
@Kyle-Verhoog I can get behind that. I considered it, but initially decided against it as this is such a niche change. Definitely happy to add that parameter though. 

Which way do you think the default should go? I was thinking it should favor the new, working trace propagation format.

**Comment by Kyle-Verhoog** [comment]
> Which way do you think the default should go? I was thinking it should favor the new, working trace propagation format.

Yup - agreed :+1:

The configuration option can be added here: https://github.com/DataDog/dd-trace-py/blob/7e1d82a399c3a2e444c14d2b65f73d738db97064/ddtrace/contrib/botocore/patch.py#L35-L41. We should offer an environment variable too

and then documented here https://github.com/DataDog/dd-trace-py/blob/7e1d82a399c3a2e444c14d2b65f73d738db97064/ddtrace/contrib/botocore/__init__.py#L16-L20

(this will show up here: https://ddtrace.readthedocs.io/en/stable/integrations.html#botocore)


Also an `upgrade` release note entry would be required here which can be created with [reno](https://docs.openstack.org/reno/latest/):

 ```shell
$ reno new botocore
```

**Comment by agocs** [comment]
Awesome! I appreciate it!

**Comment by agocs** [comment]
@Kyle-Verhoog Changes made! Thanks. Should I drop a link to this in the apm-python channel?

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]
:+1: solid tests - thanks for also keeping a test with the old setting! Just some suggestions around the user-facing aspects 🙂

File: ddtrace/contrib/botocore/patch.py [hunk 1]
```diff
@@ -37,6 +37,9 @@
     "botocore",
     {
         "distributed_tracing": get_env("botocore", "distributed_tracing", default=True),
+        "clientcontext_custom_add_datadog_object": get_env(
+            "botocore", "clientcontext_custom_add_datadog_object", default=False
```
[the remark this answers is not in the document]
Comment (agocs) [reply ddtrace/contrib/botocore/patch.py:41]: `invoke_with_legacy_context` ?

File: ddtrace/contrib/botocore/__init__.py [hunk 2]
```diff
@@ -24,6 +24,25 @@
...
+    When a tracing a directly invoked lambda function, the trace context is added to the
+    context.clientContext.custom field.
+    The new method is to add the trace context directly to the custom field, i.e.
+    ``custom{"datadog-trace-id"="12345"}``.
+    When set to True, an object named "_datadog" containing the trace context is added
+    to the context.clientContext.custom field.
+    This is to preserve legacy behavior, and is not compatible with directly invoking Go
+    or Java Lambda functions.
+
+    Please note: setting this to True is required for propagating traces to Lambda
```
[the remark this answers is not in the document]
Comment (agocs) [reply ddtrace/contrib/botocore/__init__.py:32]: 👍

File: ddtrace/contrib/botocore/__init__.py [hunk 3]
```diff
@@ -24,6 +24,25 @@
 
    Default: ``True``
 
+.. py:data:: ddtrace.config.botocore['clientcontext_custom_add_datadog_object']
+
+    When a tracing a directly invoked lambda function, the trace context is added to the
```
[the remark this answers is not in the document]
Comment (agocs) [reply ddtrace/contrib/botocore/__init__.py]: 👍

File: releasenotes/notes/botocore-46f5bf04f64777b4.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+---
+features:
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/botocore-46f5bf04f64777b4.yaml:2]: this should be an `upgrade`-level note mentioning the change and the backwards compatibility option as users using older python/node lambda client versions will have a breaking change

File: ddtrace/contrib/botocore/__init__.py [hunk 5]
```diff
@@ -24,6 +24,19 @@
 
    Default: ``True``
 
+.. py:data:: ddtrace.config.botocore['invoke_with_legacy_context']
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/botocore/__init__.py:27]: :+1: I think I'm good with this name.

@DataDog/apm-python thoughts?

File: releasenotes/notes/botocore-46f5bf04f64777b4.yaml [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+  - |
+    botocore: added `invoke_with_legacy_context` configuration setting, which is disabled by default. This configuration setting is to preserve a legacy behavior in which, when tracing a direct lambda invocation, trace context was propagated by storing it in a `_datadog` object in `context.clientContext.custom`. This must be set to True if you are invoking Lambda functions instrumented with datadog-lambda-python < v41 or datadog-lambda-js < v3.57.0. However, if set to True, invocations to Go or Java Lambda functions may fail.
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/botocore-46f5bf04f64777b4.yaml:4]: I'd probably trim this down as well. From the user standpoint they just need to know that context management support has been added for newer lambda libraries and that there is a legacy configuration option in-case they wish to run an older version.

:+1: for mentioning the specific versions of each library.

Would also be good to explicitly show the configuration options, something like:

```
Legacy support for older libraries is available with ddtrace.config.botocore.invoke_with_legacy_context = True or by setting the environment variable DD_BOTOCORE_INVOKE_WITH_LEGACY_CONTEXT=true
```

**Review by Kyle-Verhoog — COMMENTED** [review]
nice! just some last concerns about the release note but otherwise lgtm!

File: releasenotes/notes/botocore-46f5bf04f64777b4.yaml (hunk 6 above)
Comment (Kyle-Verhoog) [reply releasenotes/notes/botocore-46f5bf04f64777b4.yaml:4]: I'd like to see it explicitly mentioned that the propagation will break for users using `datadog-lambda-python < v41 or datadog-lambda-js < v3.57.0` first and then mention the introduction of the legacy setting that can be used to support these older versions.

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]
Just one more thought on the release note!

**Review by brettlangdon — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]
👏 thanks for bearing with my nitty documentation feedback @agocs!

**Pull request merged**